### ISSUE TITLE
Ctrl-C no longer quits a hanging "zq -" (regression at eb43f03)

### ISSUE THREAD

**Issue by philrz** [brimdata/zed#2993]
The most recent repro of this problem is at Zed commit 59dc184.

This journey all began with a typo. I wanted to check the version of my currently-installed `zq`, but made a typo and accidentally included an additional space character, causing `zq` to hang while waiting for input from stdin.

```
$ zq - version
```

I tried hitting Control-C multiple times to get back to my shell, but to my surprise, this did not work. I ultimately went the Control-\\ route and saw the resulting stack dump.

```
$ zq - version
^C^C^C^C^C^C^C^C^C^C^C^C^C^C^\SIGQUIT: quit
PC=0x7fff20398cde m=0 sigcode=0

goroutine 0 [idle]:
runtime.pthread_cond_wait(0x1e320c0, 0x1e32080, 0x0)
	/usr/local/opt/go/libexec/src/runtime/sys_darwin.go:384 +0x39
runtime.semasleep(0xffffffffffffffff, 0xc000048000)
	/usr/local/opt/go/libexec/src/runtime/os_darwin.go:63 +0x8d
runtime.notesleep(0x1e31e70)
	/usr/local/opt/go/libexec/src/runtime/lock_sema.go:181 +0xdb
runtime.mPark()
	/usr/local/opt/go/libexec/src/runtime/proc.go:1340 +0x39
runtime.stoplockedm()
	/usr/local/opt/go/libexec/src/runtime/proc.go:2495 +0x6e
runtime.schedule()
	/usr/local/opt/go/libexec/src/runtime/proc.go:3103 +0x48c
runtime.park_m(0xc000102a80)
	/usr/local/opt/go/libexec/src/runtime/proc.go:3318 +0x9d
runtime.mcall(0x106ed96)
	/usr/local/opt/go/libexec/src/runtime/asm_amd64.s:327 +0x5b

goroutine 1 [syscall]:
syscall.syscall(0x10b5800, 0x0, 0xc00039e000, 0x2000, 0x0, 0x0, 0x0)
	/usr/local/opt/go/libexec/src/runtime/sys_darwin.go:19 +0x2e
syscall.read(0x0, 0xc00039e000, 0x2000, 0x2000, 0x0, 0x1, 0x1)
	/usr/local/opt/go/libexec/src/syscall/zsyscall_darwin_amd64.go:1242 +0x66
...
```

This was puzzling to me, as I seemed to recall this having worked in the past. Indeed, I went on a binary search and confirmed the following.

1.  A single Control-C caused a successful exit back at the last GA release `v0.29.0`. Granted, at that time a simple `zq -` produced the warning about how it refused to write binary ZNG data to the terminal, so I had to get into the equivalent repro state with `zq -f zson -`. But a single Control-C did the trick.  
     
2.  Starting with Zed commit e2769ab (which is associated with #2596, cc: @mccanne), I started having to hit Control-C _twice_. As a user, I find this isn't too bad, since when one Control-C fails to have the desired effect, I'll usually bang on it a few times out of anger and be content if it's effective. :smile:  
     
3.  Starting with Zed commit eb43f03 (which is associated with #2919, cc: @nwt) I see the current behavior I reported above where no amount of banging Control-C makes it quit.

In addition to Control-\\, I also found hitting Control-D _three_ times also gets it to exit. I'm accustomed to Control-D being interpreted as EOF, so this was somewhat intuitive, though having to do it three times was not.

Bottom line: It's not like we have a hard requirement as to what's the ideal behavior here. However, knowing there's surely some users like me that reach for Control-C in these situations, and it used to exit with just one tap, it's probably worth seeing if we can bring back the prior behavior without breaking something else.

**Comment by philrz** [comment]
Verified in Zed commit e72fe38b.

@nwt mentioned having had to deal with lots of corner cases in this fix, but for the simple example that kicked off the issue, the attached video shows that a single Control-C now has the expected result of killing the hanging `zq` process.

https://user-images.githubusercontent.com/5934157/134590966-39efbad5-9127-4206-8767-db4773671143.mp4

Thanks @nwt!
